### ISSUE TITLE
Pagination not returning more than 50 pages

### ISSUE THREAD

**Issue by fac3-SMC** [mattmcnabb/OneLogin#52]
Hi Matt,

I wanted to ask for your input on an issue we are having. Thank you for contributing this well written Powershell library for OneLogin. We are working on a project that outputs the user data from OneLogin using this library. However, we noticed that the program encounters an error once we reach 2550 objects or 51 pages. We have 10,000 users which indicated that there was a problem reading the additional pages provided from the API. We looked at line 51-53 of "invoke-OneLoginRestMethod.ps1" and tried to alter the `$Response.data.count -eq 50` to a higher number but this did nothing it still stopped reading at 2550. 

 ```
 }
    while ($Response.Pagination.after_cursor -and $Response.data.count -eq 50)
}

```
We then removed the `-and $Response.data.count -eq 50` entirely and the program outputted all the values. Based on the code this line appears to expect 50 objects per page which is correct according to the API documentation. However, do you why this line would restrict the function from returning more than 2550 results?

**Comment by mattmcnabb** [comment]
@fac3-SMC Sorry about the issue you've encountered! I believe the line you quoted above is actually used to determine when the last page of data is returned. For instance, if a page of data contains 50 objects then we need to check the next page. If a page contains less than 50 objects then that indicates it is the last page. This is essentially a bounds check to prevent an infinite loop.

I don't recall seeing this issue before, so I'll need to dig in a bit to see if maybe something has changed in the API or if there is some case that I didn't consider.

**Comment by fac3-SMC** [comment]
@mattmcnabb Thanks for getting back so quickly. That makes perfect sense why you would do a bounds check. Im not entirely sure how to do a check to see if pagination is the issue so I will wait to hear back from you. I also did want to add that I worked with OneLogin support to debug the API calls from their end and it appears every time we query for all users the output stops at 2550. We are no where near the API call limits to rule that out.

**Comment by fac3-SMC** [comment]
@mattmcnabb hope you are having a great new year. Please let me know if there any kind of assistance we can offer to help fix this bug.

**Comment by mattmcnabb** [comment]
@fac3-SMC Sorry for the delay here - I am planning to take a deeper look at this, maybe tonight. I'll have to do some testing with an actual Onelogin tenant so I'll need to add a ton of users to my dev tenant, assuming Onelogin will allow me to do so.

**Comment by mattmcnabb** [comment]
@fac3-SMC So I'm doing a bit of testing around this, and so far I'm not experiencing the same behavior.  Here is the count of users from the bulk operations menu in OneLogin:

![image](https://user-images.githubusercontent.com/6998507/72310110-e96ff900-364e-11ea-81f2-44bed0c2753d.png)


And here is the count I'm getting from the module:

![image](https://user-images.githubusercontent.com/6998507/72310135-fbea3280-364e-11ea-9030-bebf167541e5.png)

I'm sure there's something I'm missing here. I'll need to keep digging.

**Comment by fac3-SMC** [comment]
@mattmcnabb so we performed the same exact command and returned the 2550 result again. I think the error in red might help us. This has been occurring for some reason and I have no idea why this is happening. Perhaps you may know?

![Screen Shot 2020-01-14 at 8 12 43 AM](https://user-images.githubusercontent.com/41971883/72361905-13eda080-36a7-11ea-8c3a-f75265000ab7.png)

![OneLogin_Count](https://user-images.githubusercontent.com/41971883/72362090-5fa04a00-36a7-11ea-8c06-1724ad2f3516.png)

I think this could be causing the problem?

**Comment by mattmcnabb** [comment]
That's interesting! I'm wondering under what conditions the command would output a null object to `ConvertTo-OneLoginObject`? What happens if you run the same command without `Measure-Object`? Do you get the error right at the end?

**Comment by fac3-SMC** [comment]
Yes, the same error occurs either way you run the command. The error is strange and occurs when you run the Get-OneLoginUser commandlet.

**Comment by mattmcnabb** [comment]
> Yes, the same error occurs either way you run the command. The error is strange and occurs when you run the Get-OneLoginUser commandlet.

Is the error the last thing that comes out of the pipeline?

If the convert command is what's actually hosing the pipeline and not paging, then maybe this is just a fix in error handling. That still doesn't seem to match up with your initial findings which indicated that removing the last condition in the while clause fixed your problem.

**Comment by fac3-SMC** [comment]
@mattmcnabb I apologize I don't think I told you that we did some more testing and removing that last condition fixed the problem. The error still occurred. We added it back since that did not fix it.

**Comment by fac3-SMC** [comment]
Also, to answer your last question. Yes, the error I posted is the last thing that comes out of the pipeline.

**Comment by mattmcnabb** [comment]
Ok, that helps my brain make sense of this 🤣 

Can you test a change for me?

Let's try checking to make sure that we only output the data returned from the API if it exists. In the Invoke-OneLoginRestMethod file starting at line 43, change this:

```powershell
$Response = Invoke-RestMethod @Splat -ErrorAction Stop 
$Response.Data
```

to this:

```powershell
$Response = Invoke-RestMethod @Splat -ErrorAction Stop 
if ($Response.Data) {$Response.Data}
```
That may prevent sending null to the convert command. Thanks!

**Comment by fac3-SMC** [comment]
@mattmcnabb We went ahead and made the change and still see the exact same error that I sent earlier. My apologies for confusing you. A key piece of information :upside_down_face:

**Comment by mattmcnabb** [comment]
No worries. Back to the drawing board I guess... As a temporary workaround, I may take a look at just better error handling until I can figure out why there would be a null object in the pipeline. That way you may still get an error at some point, but the script won't stop entirely.

**Comment by fac3-SMC** [comment]
I can only imagine that there is a user object that might have some sort of character that is breaking the script... When we looked at the output it always stopped at the same user but when we looked we could not see anything wrong. There was no way of us to see what the next user it was trying to process either...

**Comment by mattmcnabb** [comment]
@fac3-SMC I'm at a bit of a loss trying to imagine what might cause this error, and I'm not able to reproduce it so far. Can you try removing the `Mandatory` attribute from the `ConvertTo-OneLogin` function to see what happens?
